Pull Scripts out of PHP & Begin removing jQuery dependency

jblz opened the pull request:
This is the first pass at pulling the plugin JavaScript out of PHP files and into enqueued assets.

The [widget script](https://github.com/Parsely/wp-parsely/blob/14c342a9bca717850801fadb00aea7ae8d481761/class-parsely-recommended-widget.php#L111-L217) is not included in this change as this PR was already getting pretty large.  In addition, rewriting it with vanilla JavaScript is not a trivial effort and we're considering re-working it anyhow.

As such, this does not fully complete #134 or #158, but it's at least a step in that direction.

Additionally, we're able to use the WordPress i18n tooling to wrap a string in the Admin page JS for translation (see 3c62c8d). Again, it doesn't fully satisfy #160, but it paves the way for doing so.

## To Test

### Automated Tests

* changes should be correct and comprehensive
* Automated tests should pass

### Building the Assets

* Follow instructions to [build the included assets](https://github.com/Parsely/wp-parsely/blob/remove/jquery-dependency/DEVELOP.md#building-included-assets)
* Make sure the instructions are clear and easy to follow


### Test the plugin

* The tracking pixel should be correctly enqueued
* The API initialization script should be correctly enqueued
* The Admin JS should be correctly enqueued
* The Admin CSS should be correctly applied
* The Recommended Content widget loading should be unaffected
* Recommended content should be appropriately displayed
* In general, there should be no changes that are noticeable to end users by virtue of this change (unless they open their dev tools panel 😄)

jblz committed:
append a script tag instead of jQuery jsonp

jblz committed:
rm ensure_jquery_exists function that enqueues jquery

jblz committed:
write admin recrawl notices w/ vanilla js

jblz committed:
Pull script_loader_tag action into method

jblz committed:
build & enqueue the admin page js

jblz committed:
limit settings page script to the parsely settings page

jblz committed:
functions for asset cache buster & identifying test env

jblz committed:
Tests for script enqueues and prints, cache buster, & plugin version
* pull PARSELY_PLUGIN_URL into a constant & ref in the test
* cover multisite as well
* drop the $parsely_html file since scripts are no longer printed directly

jblz committed:
Use wp.domReady

jblz committed:
Wrap the string in the admin js for translation

jblz committed:
include built assets

jblz committed:
gitignore source map files used in dev

jblz committed:
Specify node & npm versions in package.json & .nvmrc

jblz committed:
include package-lock.json

jblz committed:
Document instructions to build included assets

jblz committed:
Pull API initialization into a js file. Enqueue it and the tracking code w/ wp functions

jblz committed:
Split out API initialization and only source when secret is set

jblz committed:
Improve script handles, only call wp_localize_script if loading api

jblz commented on src/lib/init-api/index.js:
```diff
@@ -0,0 +1,40 @@
+function uuidProfileCall( { apikey, uuid } ) {
+	const profileScript = document.createElement( 'script' );
+	profileScript.src = `https://api.parsely.com/v2/profile?apikey=${ encodeURIComponent(
+		apikey
+	) }&uuid=${ uuid || '' }&url=${ window.location.href }&callback=&_=${ +new Date() }`;
+	document.querySelector( 'head' ).appendChild( profileScript );
+}
```
This is the part I'm least confident about.  I'm attempting to replace this "jsonp" call:

https://github.com/Parsely/wp-parsely/blob/14c342a9bca717850801fadb00aea7ae8d481761/parsely-javascript.php#L24-L32

...which isn't obvious to me how it's used or what it does.  @hbbtstar -- can you shed any light on this?

GaryJones replied:
As per the internal P2 post, CORS support in browsers is now enough, that JSONP isn't needed any more.

jblz committed:
still enqueue jquery when rendering the widget (for now)

GaryJones commented on build/admin-page.asset.php:
```diff
@@ -0,0 +1 @@
+<?php return array('dependencies' => array('wp-polyfill'), 'version' => '6d8d3cc10bb4327a5d847f75b4f497e1');
```
Where does this hash come from?

jblz replied:
This is a file that's generated by `wp-scripts build` process...

Looks like it's a way to dynamically include the dependencies on the WP side:
https://developer.wordpress.org/block-editor/reference-guides/packages/packages-dependency-extraction-webpack-plugin/#wordpress

I'll give it a try to see if it's worth the "magic" to avoid having to maintain the dependency list or not.

jblz replied:
ok, magic invoked in a2131bec664d422e3e9d578da4ab7c6c297b2001.

Pretty cool.  It lets us use `import` syntax and not have to manage dependencies inside the PHP side of the plugin.

GaryJones commented on class-parsely-recommended-widget.php:
```diff
@@ -92,6 +92,7 @@ public function widget( $args, $instance ) {
 		$title = apply_filters( 'widget_title', $instance['title'] );
 
 		wp_enqueue_style( 'wp-parsely-style' );
+		wp_enqueue_script( 'jquery' );
```
Instead of enqueueing jQuery directly, can you add it as a dependency arg for the registration/enqueue of the now-external .js files?

jblz replied:
> Instead of enqueueing jQuery directly, can you add it as a dependency arg for the registration/enqueue of the now-external .js files?

The Widget script is not yet extracted from the PHP, so it has to be explicitly enqueued here.

Once we do pull it out, we'll definitely want to add it as a dependency to script (unless we remove the dependency entirely, of course).

GaryJones commented on package.json:
```diff
@@ -6,9 +6,13 @@
     "test": "tests"
   },
+  "engines": {
+    "node": "^14.16.1",
+    "npm": "^7.10.0"
```
These seem like really high dependency versions. What is it about these specific versions that are required to make the build work? What would the lowest versions of node and npm be that would still work?

jblz replied:
> These seem like really high dependency versions. What is it about these specific versions that are required to make the build work? What would the lowest versions of node and npm be that would still work?

These are the latest versions of the LTS node release & the accompanying `latest` tag of npm. While there's nothing tying it to these particular versions functionality-wise, the engine is only developer-facing and it seems reasonable to enforce compatibility with the current point releases.

GaryJones replied:
Since it's the latest LTS and latest (minor) release, could it be:

```suggestion
    "node": "^14",
    "npm": "^7"
```

...instead? That gives a bit of wiggle room for contributors to start contributing and building, even if they aren't on the latest minor / patch version.

Right now, the latest npm is 7.11.1.

GaryJones commented on src/lib/admin-page/index.js:
```diff
@@ -0,0 +1,19 @@
+	}
+
+	const notice = sprintf(
+		'<p class="description"><strong style="color:red;">Important:</strong> changing this value on a site currently tracked with Parse.ly will require reprocessing of your Parse.ly data. Once you have changed this value, please contact <a href="mailto:support@parsely.com?subject=Please reprocess %s">support@parsely.com</a></p>',
```
```suggestion
		'<p class="description"><strong style="color:#d63638;">Important:</strong> changing this value on a site currently tracked with Parse.ly will require reprocessing of your Parse.ly data. Once you have changed this value, please contact <a href="mailto:support@parsely.com?subject=Please reprocess %s">support@parsely.com</a></p>',
```

Just changes the red slightly to match the notice/form error red colour that is in the default admin colour scheme (aiming for consistency). Here's a comparison (new red is second instance):
<img width="325" alt="Screenshot 2021-04-23 at 10 14 43" src="https://user-images.githubusercontent.com/88371/115849674-059b4200-a41d-11eb-9a7e-29002810de7a.png">

Alternatively, leave as `red` for now, and we can reassess this when it's refactored to remove the noise (since technically this "Important" isn't an error).

jblz replied:
I'm definitely good with this change, but I'd pull it out to a separate PR.

The color value is coming from here:
https://github.com/Parsely/wp-parsely/blob/1a398fac2ebc4935199299977522c5a817b62505/parsely-admin-header.php#L32

GaryJones replied:
Opened https://github.com/Parsely/wp-parsely/issues/254.

GaryJones commented on tests/all-test.php:
```diff
@@ -103,30 +101,113 @@ public function setUp() {
 	}
 
+	public function test_class_version() {
+		self::assertEquals( '2.4.1', \Parsely::VERSION );
```
This seems fragile, as in it will be another thing to update on each release.

The only thing it seems to be testing is that someone can write a 5-character string correctly in two places.

jblz replied:
The reason I added the explicit test of the value of the `\Parsely::VERSION` class constant is that it is being used in the cache buster by virtue of the `PARSELY_VERSION` defined global constant.

If it makes sense to cover the behavior around the functionality that sets the cache buster (& I contend that it is), it makes sense to go the extra mile with the other two values with the single hard-coded version string.

When we bump the version in the plugin, the test will fail if this line is missed & it's an easy change.

GaryJones replied:
> If it makes sense to cover the behavior around the functionality that sets the cache buster (& I contend that it is), it makes sense to go the extra mile with the other two values with the single hard-coded version string.

For clarity, when we're talking about the cache-busting, we mean setting the arg in the `wp_register_script()` / `wp_enqueue_script()`?

Ensuring that arg is set in one of those function calls, is ours to test against. Checking that the value is used in the `<script>` output, is not as this is WP code that adds the querystring.

Happy to let it go in, but we'll want to review how it fits in when we audit and cleanup `all-tests.php`.

GaryJones commented on tests/all-test.php:
```diff
@@ -103,30 +101,113 @@ public function setUp() {
+		echo self::$parsely->insert_parsely_javascript();
+
+		$intermediate_output = ob_get_contents();
+		self::assertEquals(
```
```suggestion
		self::assertSame(
```

Does a type check (`===`) as well, instead of just the value check (`==`).

GaryJones commented on tests/all-test.php:
```diff
@@ -103,30 +101,113 @@ public function setUp() {
-		echo esc_html( $output );
-		self::assertContains( self::$parsely_html, $output );
+
+		self::assertEquals(
```
```suggestion
		self::assertSame(
```

GaryJones commented on tests/all-test.php:
```diff
@@ -103,30 +101,113 @@ public function setUp() {
-		self::assertContains( self::$parsely_html, $output );
+
+		self::assertEquals(
+			"<script type='text/javascript' data-parsely-site=\"blog.parsely.com\" src='https://cdn.parsely.com/keys/http://blog.parsely.com/p.js?ver=" . PARSELY_VERSION . "' id=\"parsely-cfg\"></script>\n",
```
Aside: that type='text/javascript' is redundant for HTML5, since that's the default value.

GaryJones replied:
Opened https://github.com/Parsely/wp-parsely/issues/255.

GaryJones commented on tests/all-test.php:
```diff
@@ -103,30 +101,113 @@ public function setUp() {
+	 * Test the parsely script enqueues when an api secret is set.
+	 */
+	public function test_parsely_api_enabled_scripts() {
+		$options = get_option( 'parsely' );
+		$options['api_secret'] = 'hunter2';
+		update_option( 'parsely', $options );
```
Might want to consider moving this utility step into its own method:

```php
private set_api_secret( $api_secret ) {
	$options = get_option( 'parsely' );
	$options['api_secret'] = $api_secret;
	update_option( 'parsely', $options );
}
```

Then:
```suggestion
		$this->set_api_password( 'hunter2' );
```

GaryJones replied:
Or if the other tests set other settings, then make it slightly more generic:

```php
private update_parsely_option( $key, $value ) {
	$options = get_option( 'parsely' );
	$options[ $key ] = $value;
	update_option( 'parsely', $options );
}

GaryJones commented on tests/all-test.php:
```diff
@@ -103,30 +101,113 @@ public function setUp() {
+		echo self::$parsely->insert_parsely_javascript();
+
+		$intermediate_output = ob_get_contents();
+		self::assertEquals(
```
```suggestion
		self::assertSame(
```

GaryJones commented on tests/all-test.php:
```diff
@@ -103,30 +101,113 @@ public function setUp() {
+		$post_array = $this->create_test_post_array();
+		$post       = $this->factory->post->create( $post_array );
+		$this->go_to( '/?p=' . $post );
+		echo self::$parsely->insert_parsely_javascript();
```
Might be beyond the scope of this PR, but instead of echoing out the value into an output buffer, can we adjust the function to just get a return and check against that instead? #180.

GaryJones commented on tests/all-test.php:
```diff
@@ -459,17 +575,79 @@ public function test_user_logged_in_multisite() {
 		$output = ob_get_clean();
-		self::assertContains( self::$parsely_html, $output );
+
+		self::assertEquals(
```
```suggestion
		self::assertSame(
```

GaryJones commented on tests/all-test.php:
```diff
@@ -459,17 +575,79 @@ public function test_user_logged_in_multisite() {
+		echo self::$parsely->insert_parsely_javascript();
+
+		$intermediate_output = ob_get_contents();
+		self::assertEquals(
```
```suggestion
		self::assertSame(
```

GaryJones commented on tests/all-test.php:
```diff
@@ -459,17 +575,79 @@ public function test_user_logged_in_multisite() {
 		$output = ob_get_clean();
-		self::assertNotContains( self::$parsely_html, $output );
+
+		self::assertEquals(
```
```suggestion
		self::assertSame(
```

GaryJones commented on tests/all-test.php:
```diff
@@ -459,17 +575,79 @@ public function test_user_logged_in_multisite() {
+		echo self::$parsely->insert_parsely_javascript();
+
+		$intermediate_output = ob_get_contents();
+		self::assertEquals(
```
```suggestion
		self::assertSame(
```

GaryJones commented on tests/all-test.php:
```diff
@@ -419,10 +500,45 @@ public function test_user_logged_in() {
 		$output = ob_get_clean();
-		self::assertNotContains( self::$parsely_html, $output );
+
+		self::assertEquals(
```
```suggestion
		self::assertSame(
```

GaryJones commented on tests/all-test.php:
```diff
@@ -419,10 +500,45 @@ public function test_user_logged_in() {
+		echo self::$parsely->insert_parsely_javascript();
+
+		$intermediate_output = ob_get_contents();
+		self::assertEquals(
```
```suggestion
		self::assertSame(
```

GaryJones commented on tests/all-test.php:
```diff
@@ -103,30 +101,113 @@ public function setUp() {
+		wp_print_scripts();
+		$output = ob_get_clean();
+
+		self::assertEquals(
```
```suggestion
		self::assertSame(
```

GaryJones commented on wp-parsely.php:
```diff
@@ -104,12 +104,15 @@ public function __construct() {
 		add_action( 'instant_articles_compat_registry_analytics', array( $this, 'insert_parsely_tracking_fbia' ) );
 		add_action( 'template_redirect', array( $this, 'parsely_add_amp_actions' ) );
-		if ( ! defined( 'WP_PARSELY_TESTING' ) ) {
+		if ( ! self::is_test_env() ) {
```
Outside of the scope of this PR, but this really shouldn't be needed! Need to see why tests supposedly break when some styles are enqueued!

GaryJones replied:
Opened https://github.com/Parsely/wp-parsely/issues/256.

GaryJones commented on wp-parsely.php:
```diff
@@ -1046,15 +1071,65 @@ public function insert_parsely_javascript() {
+
+		wp_enqueue_script(
+			'wp-parsely-tracker',
+			'https://cdn.parsely.com/keys/' . esc_url( $parsely_options['apikey'] ) . '/p.js',
```
Escaping of a URL is handled by WP when it outputs the HTML (late), so it's not needed here.

GaryJones commented on wp-parsely.php:
```diff
@@ -1046,15 +1071,65 @@ public function insert_parsely_javascript() {
+		}
+
+		if ( $handle === 'wp-parsely-tracker' ) {
+			$tag = preg_replace( '/ id=(\"|\')wp-parsely-tracker-js\1/', ' id="parsely-cfg"', $tag );
```
Do we know why this change of ID is needed?

jblz replied:
I really don't know. This is another case of trying to not break anything without understanding why it is the way it is.

@hbbtstar -- any thoughts as to the necessity of having particular `id` attributes to the `script` tags?

GaryJones reviewed (changes requested):
I love where this is going!

Added a few comments, including 3-4 items that we may want to spin up into separate issues/PRs, if you don't want to address them in this one.

jblz committed:
Use assertSame instead of assertEquals

Co-authored-by: Gary Jones <gary.jones@automattic.com>

jblz committed:
Do not call esc_url on the apikey value in the enqueue call

jblz committed:
Import dependencies & dynamically include them in WordPress deps

jblz committed:
pin engine settings to "compatible w/ node 14 & npm 7"

Co-authored-by: Gary Jones <gary.jones@automattic.com>

jblz committed:
Merge branch 'develop' of github.com:Parsely/wp-parsely into remove/jquery-dependency

jblz committed:
Remove element specificity for the recrawl message per #253

jblz committed:
update admin page built assets

jblz committed:
Merge branch 'remove/jquery-dependency' of github.com:Parsely/wp-parsely into remove/jquery-dependency

GaryJones commented:
Looks like the `parsely-admin-header.php` file can be removed as well, as the `include` to it is being removed.

GaryJones committed:
Remove custom test constant

Existing tests pass with and without this constant, which means it's not doing anything (which is a good thing).

The only thing that changed is the cache busting string changed when WP_DEBUG was defined (not also checks it is truthy), so we can use an existing WP test constant for that, instead of defining our own.

Fixes incorrectly named static variable, adds documentation to the filter.

Fixes #256.

jblz committed:
Move top-level PHP files to src per #262

jblz committed:
Merge branch 'develop' of github.com:Parsely/wp-parsely into remove/jquery-dependency

jblz committed:
update package-lock

jblz committed:
Move js files to src/js

jblz committed:
Rm unused src/parsely-admin-header.php

jblz committed:
fixup merged class-all-tests.php

jblz committed:
Define PARSELY_PLUGIN_DIR & use it in includes

jblz committed:
fixup settings include

jblz committed:
New Test Helper Parsely\Tests\TestUtils::set_options

jblz committed:
Move DEFAULT_OPTIONS to a class constant

GaryJones commented on tests/class-testutils.php:
```diff
@@ -0,0 +1,20 @@
+namespace Parsely\Tests;
+
+class TestUtils {
+	public static function set_options( $custom_options = array() ) {
```
We already have a custom TestCase that utility methods should go into.

jblz committed:
Move the set_options helper to the existing TestCase class

jblz committed:
Convert Profile API call to simple XHR - no more JSONP

jblz committed:
build assets

jblz committed:
Merge pull request #273 from Parsely/remove/jquery-dependency--fetch-instead-of-jsonp

Convert Profile API call to simple XHR - no more JSONP

jblz committed:
Merge branch 'develop' of github.com:Parsely/wp-parsely into remove/jquery-dependency

jblz committed:
Fix Translation Wrapping of the Recrawl Notice

jblz committed:
include updated built assets

jblz committed:
Remove src/parsely-parsely-page.php -- it's handled in the views files now

jblz committed:
Prevent inclusion of package-lock from the built plugin

GaryJones commented on src/class-parsely.php:
```diff
@@ -1069,6 +1073,22 @@ function( $v ) {
+		$cache_buster = defined( 'WP_DEBUG' ) && WP_DEBUG && empty( 'WP_TESTS_DOMAIN' ) ? wp_rand() : PARSELY_VERSION;
+
+		/**
+		 * Filters the cache buster value for linked scripts and styles.
+		 *
+		 * @param string $cache_buster Plugin version, unless WP_DEBUG is defined and truthy, and tests are not running.
```
```suggestion
		 * Filters the cache buster value for linked scripts and styles.
		 *
		 * @since 2.5.0
		 *
		 * @param string $cache_buster Plugin version, unless WP_DEBUG is defined and truthy, and tests are not running.
```

GaryJones commented on src/js/lib/admin-page/index.js:
```diff
@@ -0,0 +1,24 @@
+	}
+
+	const notice = sprintf(
+		// translators: %s: The API Key that will be used to request a "recrawl"
```
```suggestion
		/* translators: %s: The API Key that will be used to request a recrawl */
```

GaryJones committed:
JS: Follow standard translator comment format

GaryJones committed:
Filter: Add since tag
